cmake/modules/dts.cmake

```diff
@@ -69,246 +141,362 @@ foreach(dts_dir ${DTS_ROOT})
-  # each line actually came from, which improves error reporting.
+#
+# If we have a system devicetree, we need to generate SYSDT_OUTPUT_DTS
+# before we can handle the "normal" devicetree.
+#
+
+if(BOARD_HAS_SYSDTS)
+  # Run the preprocessor on the input system devicetree files. We are
+  # leaving linemarker directives enabled on purpose. This tells
+  # sysdtlib where each line actually came from, which improves error
+  # reporting.
   execute_process(
     COMMAND ${CMAKE_DTS_PREPROCESSOR}
     -x assembler-with-cpp
     -nostdinc
     ${DTS_ROOT_SYSTEM_INCLUDE_DIRS}
-    ${DTC_INCLUDE_FLAG_FOR_DTS}  # include the DTS source and overlays
+    -include ${BOARD_SYSDTS} # Include BOARD.sysdts.
     ${NOSYSDEF_CFLAG}
     -D__DTS__
     ${DTS_EXTRA_CPPFLAGS}
     -E   # Stop after preprocessing
     -MD  # Generate a dependency file as a side-effect
-    -MF ${DTS_DEPS}
-    -o ${DTS_POST_CPP}
+    -MF ${SYSDT_DEPS}
+    -o ${SYSDT_POST_CPP}
     ${ZEPHYR_BASE}/misc/empty_file.c
     WORKING_DIRECTORY ${APPLICATION_SOURCE_DIR}
     RESULT_VARIABLE ret
     )
-  if(NOT "${ret}" STREQUAL "0")
-    message(FATAL_ERROR "command failed with return code: ${ret}")
+  if(NOT ret EQUAL 0)
+    message(FATAL_ERROR "system devicetree preprocessing failed with return code: ${ret}")
   endif()
 
-  # Parse the generated dependency file to find the DT sources that
-  # were included, including any transitive includes, and then add
-  # them to the list of files that trigger a re-run of CMake.
-  toolchain_parse_make_rule(${DTS_DEPS}
-    include_files # Output parameter
-    )
+  # If we have a system devicetree, the user must specify the path to the
+  # execution domain that they want to select via SYSDT_EXECUTION_DOMAIN.
+  if(NOT DEFINED SYSDT_EXECUTION_DOMAIN)
+    message(FATAL_ERROR "Found system devicetree file ${BOARD_SYSDTS}, but SYSDT_EXECUTION_DOMAIN was not defined. Set this variable to the path to the execution domain you wish to use.")
+  endif()
 
-  set_property(DIRECTORY APPEND PROPERTY
-    CMAKE_CONFIGURE_DEPENDS
-    ${include_files}
-    ${GEN_DEFINES_SCRIPT}
-    ${GEN_DRIVER_KCONFIG_SCRIPT}
-    ${DTS_CMAKE_SCRIPT}
+  execute_process(COMMAND ${PYTHON_EXECUTABLE} ${GEN_DTSI_FROM_SYSDTS_SCRIPT}
+    --sysdts "${SYSDT_POST_CPP}"
+    --domain "${SYSDT_EXECUTION_DOMAIN}"
+    --sysdt-pickle-out "${SYSDT_PICKLE}"
+    --dts-out "${SYSDT_OUTPUT_DTS}"
+    RESULT_VARIABLE ret
     )
+  if(NOT ret EQUAL 0)
+    message(FATAL_ERROR "Failed to generate .dtsi from system devicetree: ${gen_dt_from_sysdt_result}")
+  endif()
+endif(BOARD_HAS_SYSDTS)
 
+#
+# Halt execution early if there is no devicetree or system devicetree.
+#
+
+if(BOARD_HAS_SYSDTS)
+  # When a system devicetree is present, we'll use the output devicetree
+  # in the build directory as DTS_SOURCE.
   #
-  # Run gen_defines.py to create a header file, zephyr.dts, and edt.pickle.
-  #
+  # TODO: how can we support board revisions in system devicetree?
```

Comment:
If the revision is still provided with the board name, can it not check if `<board>@<revision>.overlay` exists and if so, include that with the main dts file?

Reply:
So far we are trying to be careful to use different extensions for system DT and standard DT files:

- `.sysdts` vs `.dts`
- `.sysdtsi` vs `.dtsi`

So I guess this would be `.sysoverlay`?